### PR TITLE
[FLINK-11069] [utils] Remove FutureUtil

### PR DESCRIPTION
**Pull request by TisonKun**
## What is the purpose of the change

The utility class `FutureUtil` contains duplicate methods and method which are not implemented optimally. I suggest to merge `FutureUtil` with `FutureUtils` and get rid of the `waitForAll` method.

## Brief change log

- Remove `FutureUtil`
- Merge `runIfNotDoneAndGet` into `FutureUtils`
- Get rid of `waitForAll`. Instead use `FutureUtils#waitForAll`
- Introduce `uncheck` static method in `SupplierWithException` for the adjustment

## Verifying this change

This change is a trivial rework covering by the relevant tests.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (**no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (**no**)
  - The serializers: (**no**)
  - The runtime per-record code paths (performance sensitive):(**no**)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (**no**)
  - The S3 file system connector: (**no**)

## Documentation

  - Does this pull request introduce a new feature? (**no**)
  - If yes, how is the feature documented? (**not applicable**)

cc @tillrohrmann

### CODE REVIEW

**Comment by yanghua** [comment]
Hi @TisonKun 

```
18:26:28.316 [ERROR] Failed to execute goal com.github.siom79.japicmp:japicmp-maven-plugin:0.11.0:cmp (default) on project flink-core: Breaking the build because there is at least one incompatibility: org.apache.flink.util.function.SupplierWithException.unchecked(org.apache.flink.util.function.SupplierWithException):METHOD_ADDED_TO_INTERFACE -> [Help 1]
```

**Comment by TisonKun** [comment]
Thanks @yanghua.
I notice the inconsistent among `util.function`. With 1e6a91a3bf7e734eccdb034ce505b3775b709265 @StephanEwen mark the functional interfaces in `util.function` as `@Public`, but the following interfaces were not marked. Such a `unchecked` static method is proposed by @GJL.
I prefer the idea to make the interfaces final, and thus think it could be a follow up to refactor the static methods into `FunctionUtils`. Also remove `CheckedSupplier`.

File: flink-runtime/src/test/java/org/apache/flink/runtime/io/network/partition/consumer/LocalInputChannelTest.java [hunk 1]
```diff
@@ -158,27 +160,31 @@ public void testConcurrentConsumeMultiplePartitions() throws Exception {
...
-						i,
-						parallelism,
-						numberOfBuffersPerChannel,
-						networkBuffers.createBufferPool(parallelism, parallelism),
-						partitionManager,
-						new TaskEventDispatcher(),
-						partitionIds)));
+				results.add(CompletableFuture.supplyAsync(
+					CheckedSupplier.unchecked(
+						new TestLocalInputChannelConsumer(
```
Comment (zentol) [inline flink-runtime/src/test/java/org/apache/flink/runtime/io/network/partition/consumer/LocalInputChannelTest.java]: Please introduce a dedicated variable for the consumer to make this code more readable, i.e.
```
final TestLocalInputChannelConsumer consumer = ....
results.add(CompletableFuture. .... (consumer::call)...);
```

Comment (TisonKun) [reply flink-runtime/src/test/java/org/apache/flink/runtime/io/network/partition/consumer/LocalInputChannelTest.java]: done

**Pull request merged**